wen-restart/src/wen_restart.rs

```diff
@@ -198,7 +235,60 @@ pub(crate) fn aggregate_restart_last_voted_fork_slots(
             sleep(Duration::from_millis(time_left));
         }
     }
-    Ok(())
+    Ok(last_voted_fork_slots_aggregate.get_final_result())
+}
+
+// Verify that all blocks with at least (active_stake_percnet - 38%) of the stake form a
+// single chain from the root, and use the highest slot in the blocks as the heaviest fork.
+// Please see SIMD 46 "gossip current heaviest fork" for correctness proof.
+pub(crate) fn find_heaviest_fork(
+    aggregate_final_result: LastVotedForkSlotsFinalResult,
+    bank_forks: Arc<RwLock<BankForks>>,
+    blockstore: Arc<Blockstore>,
+    exit: Arc<AtomicBool>,
+) -> Result<(Slot, Hash)> {
+    // Because everything else is stopped, it's okay to grab a big lock on bank_forks.
+    let my_bank_forks = bank_forks.read().unwrap();
+    let root_bank = my_bank_forks.root_bank().clone();
+    let root_slot = root_bank.slot();
+    // TODO: Should use better epoch_stakes later.
+    let epoch_stake = root_bank.epoch_stakes(root_bank.epoch()).unwrap();
+    let total_stake = epoch_stake.total_stake();
+    let stake_threshold = aggregate_final_result
+        .total_active_stake
+        .saturating_sub((HEAVIEST_FORK_THRESHOLD_DELTA * total_stake as f64) as u64);
+    let mut slots = aggregate_final_result
+        .slots_stake_map
+        .iter()
+        .filter(|(slot, stake)| **slot > root_slot && **stake > stake_threshold)
+        .map(|(slot, _)| *slot)
+        .collect::<Vec<Slot>>();
+    slots.sort();
+    let mut expected_parent = root_slot;
+    for slot in slots {
+        if exit.load(Ordering::Relaxed) {
+            return Err(WenRestartError::Exiting.into());
+        }
+        if let Ok(Some(block_meta)) = blockstore.meta(slot) {
+            info!("block meta: {:?}", block_meta);
+            if block_meta.parent_slot != Some(expected_parent) {
```

Comment:
Hmmm if you start with a really old root `N` and everyone else's root is at `N + 1000`, are you guaranteed to have gotten all the slots between your root and everyone else's root through the aggregation phase? 

Seems like it's not guaranteed if people didn't gossip the older slots in their last voted forks

Reply:
We always send back (last_vote - 64k, last_vote), it may include the local root. So yes you should see N to N+1000 in the RestartLastVotedForkSlots and you will aggregate those.

The number of slots may be less than 64k, depending on how much we can fit in one UDP packet, but it's definitely more than 1000.

Reply:
Mayeb a custom error message would be good here if the only mismatch is on the root? Say that this means potentially we don't have enough ancestors or your root is too old?

Reply:
Sure, I think an error log on first mismatch being on root is useful. But not sure we need the error log when the root mismatch is the only error, because even if the list is screwed up we might want to print root mismatch, changed.